### ISSUE TITLE
Performance regression in LagrangesMethod

### ISSUE THREAD

**Issue by moorepants** [sympy/sympy#14674]
Offending PRs:

- https://github.com/sympy/sympy/pull/13892
- https://github.com/sympy/sympy/pull/13751
- https://github.com/sympy/sympy/pull/13641

Benchmark:

http://www.moorepants.info/misc/sympy-asv/#physics.mechanics.lagrange.LagrangesMethodMassSpringDamper.time_lagrangesmethod_mass_spring_damper

**Comment by asmeurer** [comment]
I think derivatives are just a bit slower because there's now more indirection, i.e., more function calls in the API. I don't know if there are ways they could be sped up.

**Comment by asmeurer** [comment]
This particular benchmark runs quite fast (unusually so for mechanics problems, from my understanding). Would it be possible to parameterize the benchmark so that we can more easily see if slowdowns like this scale with the problem or if they are constant?

**Comment by moorepants** [comment]
Yes I can parameterize it. @bjodah and I have been trying to keep the benchmarks running fast, since it takes so long to run the total of them. I'll look into it.

**Comment by asmeurer** [comment]
I think by default we would still just test the one variant. But if there is a parameter we can modify it to test how regressions like this scale. There are other benchmarks in the suite that have parameters that are only currently benchmarked against one value.

**Comment by certik** [comment]
Is the SymEngine backend also slower? If we want performance, I recommend to use SymEngine.

**Comment by moorepants** [comment]
We do not have the benchmark system setup to test the symengine backend.

**Comment by oscarbenjamin** [comment]
The benchmark is:
```python
from sympy import symbols
import sympy.physics.mechanics as me

# System state variables
q = me.dynamicsymbols('q')
qd = me.dynamicsymbols('q', 1)

# Other system variables
m, k, b = symbols('m k b')

# Set up the reference frame
N = me.ReferenceFrame('N')

# Set up the points and particles
P = me.Point('P')
P.set_vel(N, qd * N.x)

Pa = me.Particle('Pa', P, m)

# Define the potential energy and create the Lagrangian
Pa.potential_energy = k * q**2 / 2.0
L = me.Lagrangian(N, Pa)

# Create the list of forces acting on the system
fl = [(P, -b * qd * N.x)]

# Create an instance of Lagranges Method
l = me.LagrangesMethod(L, [q], forcelist=fl, frame=N)

# Create the equations of motion using lagranges method
for n in range(3):
    import time
    t = time.time()
    l.form_lagranges_equations()
    print('%.3g seconds' % (time.time() - t))
```
Running that on master I get
```console
$ python t.py 
0.0395 seconds
0.0281 seconds
0.0321 seconds
```
whereas sympy 1.1 gives
```
$ python t.py 
0.0146 seconds
0.00207 seconds
0.002 seconds
```
so it's become 2-3 times slower in the cold-cache case.

I've bisected the main slowdown that I see to 03ffa11604bcd59a75e3bb6acebe9e34416f8c65 from #15049 although I guess there could be earlier slowdowns (since that PR was opened after this issue).

I think that there have been many modifications to the way derivatives are handled since this issue was opened. Many things have been changed and improved, made more correct, etc. Clearly performance of many things can be improved in SymPy but I'm not sure that the slowdown reported here is worth keeping an open issue for.

Certainly this shouldn't hold up the release of 1.5 if no one is proposing to address it so I'll remove the milestone.

**Comment by moorepants** [comment]
I think that is reasonable to desire that changes that correct results and add features to issues unrelated to the benchmark cases should not significantly affect the performance. In this case, we know that this benchmark obtained the correct result in a much faster way before unrelated changes were introduced. In the same way that we don't allow changes to break existing tests, I think that trying to ensure changes don't degrade performance in existing benchmarks is a good thing to strive for. It is just unclear at how we objectify the balance between correcting other parts of the code and performance of existing code.

I think these issues should remain open for two reasons:

- someone can address this issue and likely bring back the performance
- it is important that sympy developers are aware that performance is important.

For complex expressions that are generated in this module, it may take 5 minutes to compute before this regression. Now if it takes 15 minutes, that's a big deal.

**Comment by oscarbenjamin** [comment]
> For complex expressions that are generated in this module, it may take 5 minutes to compute before this regression. Now if it takes 15 minutes, that's a big deal.

I agree. The discussion above has highlighted the need for a more comprehensive benchmark for this issue though.

Having just worked through a few of these benchmark issues I'm not entirely convinced that they all represent significant performance issues. For example looking at #14673 it seems that the issue is the one-time cost of importing the units module. That won't scale up to larger problems since it is a one-time cost. The particular benchmark in this issue completes relatively quickly so it also isn't clear that the particular slowdown reported here would be significant in a scaled up problem.

The reason I am working through these issues is to clear the way for releasing 1.5. There is a big difference between keeping an issue open and holding up a release which is why I have removed the milestone here but not closed the issue. I do think that performance considerations are important - perhaps the benchmarks would be more useful if we had a way of making them testable before a PR is merged though (I guess a topic to be discussed somewhere else...).

**Comment by moorepants** [comment]
> Having just worked through a few of these benchmark issues I'm not entirely convinced that they all represent significant performance issues.

Yes, I've only reported them if they happen to show issues in:

http://www.moorepants.info/misc/sympy-asv/

Most I haven't looked into in detail.

> I do think that performance considerations are important - perhaps the benchmarks would be more useful if we had a way of making them testable before a PR is merged though (I guess a topic to be discussed somewhere else...).

It is. We want to run benchmarks, but we need a dedicated server to do so that can't vary in terms of hardware differences.

**Comment by bjodah** [comment]
If one only compares two commits (`HEAD` & `master`) the time required to run the benchmarks on Travis should be considerably less than the time required by the test suite, right?

So why not run `asv compare --factor 1.5 HEAD master` as the final step of CI for a PR? (perhaps optionally disabled by a regex against the commit message to enable forcefully accepting a slowdown?)

**Comment by oscarbenjamin** [comment]
I think that an allowed failure on Travis could still be useful. When reviewing a PR you could highlight the particular slowdown that is reported. It wouldn't need to prevent merging.

**Comment by moorepants** [comment]
@bjodah Yes, that seems reasonable.

**Comment by moorepants** [comment]
I opened this: https://github.com/sympy/sympy/issues/17807
